cmd/boulder-janitor/job.go

```diff
@@ -0,0 +1,181 @@
+		},
+		[]string{"table"})
+	// workStat is a prometheus gauge vector tracking the number of rows found
+	// during a batchedJob's getWork stage and queued into the work channel sliced
+	// by a table label.
+	workStat = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Name: "janitor_workbatch",
+			Help: "Number of items of work by table the boulder-janitor queued for deletion.",
+		},
+		[]string{"table"})
+)
+
+// batchedDBJob is a struct abstracting the common properties of a long running
+// cleanup job based on cursoring across a database table's auto incrementing
+// primary key.
+type batchedDBJob struct {
+	db  janitorDB
+	log blog.Logger
+	// table is the name of the table that this job cleans up.
+	table string
+	// purgeBefore indicates the cut-off for the the resoruce being cleaned up by
+	// the job. Rows that older than now - purgeBefore are deleted.
+	purgeBefore time.Time
+	// batchSize indicates how many database rows of work should be returned per query.
+	batchSize int64
+	// maxDPS optionally indicates a maximum rate of deletes to run per second.
+	maxDPS int
+	// parallelism controls how many independent go routines will be performing
+	// cleanup deletes.
+	parallelism int
+	// workQuery is the parameterized SQL query that is used to find more work. It will be provided three parameters:
+	//   * :startID - the primary key value to start the work query from.
+	//   * :cutoff  - the purgeBefore date used to control which rows are old enough to be deleted.
+	//   * :limit   - the batchSize value. Only this many rows should be returned by the query.
+	workQuery string
+}
+
+// getWork reads work into the provided work channel starting at the startID by
+// using the batchedDBJob's configured work query, purgeBefore, and batchSize.
+// If there is no error the last primary key ID written to the work channel will
+// be returned, otherwise an error result is returned.
+func (j batchedDBJob) getWork(work chan<- int64, startID int64) (int64, error) {
+	var idBatch []int64
+	_, err := j.db.Select(
+		&idBatch,
+		j.workQuery,
+		map[string]interface{}{
+			"startID": startID,
+			"cutoff":  j.purgeBefore,
+			"limit":   j.batchSize,
+		},
+	)
+	if err != nil && err != sql.ErrNoRows {
+		return 0, err
+	}
+	rowsFound := len(idBatch)
+	workStat.WithLabelValues(j.table).Set(float64(rowsFound))
+	if rowsFound == 0 {
+		return startID, nil
+	}
+	var lastID int64
+	for _, v := range idBatch {
+		work <- v
+		lastID = v
+	}
+	return lastID, nil
+}
+
+// cleanResource uses the configured level of parallelism to run go routines
+// that read ID values from the work channel and delete the corresponding table
+// rows. If the batchedDBJob configures a maxDPS rate then it will be enforced by
+// synchronizing the delete operations on a ticker based on the maxDPS.
+// cleanResource will block until all of the worker go routines complete.
+func (j batchedDBJob) cleanResource(work <-chan int64) {
+	wg := new(sync.WaitGroup)
```

Comment:
Just to clarify this in my mind: is the usage of a WaitGroup here only required for testing? As far as I can tell `work` is never closed during normal operation.

Reply:
Yup, just for the unit tests.